File: atomic-x/ios/Classes/AudioPlayer/AudioPlayer.swift
import AVFoundation
import AVKit
import SwiftUI

public class AudioPlayer: NSObject, ObservableObject {
    static let shared = AudioPlayerImpl()
    
    public var onComplete: (() -> Void)?
    public var onProgressUpdate: ((_ currentPosition: Int, _ duration: Int) -> Void)?
    public var onPlay: (() -> Void)?
    public var onPause: (() -> Void)?
    public var onResume: (() -> Void)?
    public var onError: ((_ errorMessage: String) -> Void)?
    
    public func play(filePath: String) {}
    
    public func pause() {}
    
    public func resume() {}
    
    public func stop() {}
    
    public func getCurrentPosition() -> Int { return 0 }
    
    public func getDuration() -> Int { return 0 }
    
    public func isPlaying() -> Bool { return false }
    
    public func isPaused() -> Bool { return false }
    
    override internal init() {
        super.init()
    }
}

File: atomic-x/android/src/main/kotlin/io/trtc/tuikit/atomicx/audioplayer/AudioPlayer.kt
package io.trtc.tuikit.atomicx.audioplayer

import io.trtc.tuikit.atomicx.audioplayer.impl.MediaAudioPlayer

abstract class AudioPlayer {
    companion object {
        val shared: AudioPlayer by lazy {
            MediaAudioPlayer()
        }
    }

    abstract fun play(filePath: String)

    abstract fun pause()

    abstract fun resume()

    abstract fun stop()

    abstract fun setListener(listener: AudioPlayerListener?): AudioPlayer

    abstract fun getCurrentPosition(): Int

    abstract fun getDuration(): Int

    abstract fun isPlaying(): Boolean

    abstract fun isPaused(): Boolean
}

enum class AudioOutputDevice {
    SPEAKER,
    EARPIECE
}

interface AudioPlayerListener {

    fun onPlay() {}

    fun onPause() {}

    fun onResume() {}

    fun onProgressUpdate(currentPosition: Int, duration: Int) {}

    fun onCompletion() {}

    fun onError(errorMessage: String) {}

    fun onAudioOutputChanged(device: AudioOutputDevice) {}
}
